service/src/intTest/java/uk/nhs/adaptors/gp2gp/gpc/GpcWebClientTest.java

```diff
@@ -0,0 +1,226 @@
+import org.springframework.boot.test.mock.mockito.SpyBean;
+import org.springframework.test.annotation.DirtiesContext;
+import org.springframework.test.context.junit.jupiter.SpringExtension;
+
+import okhttp3.mockwebserver.MockResponse;
+import okhttp3.mockwebserver.MockWebServer;
+import okhttp3.mockwebserver.SocketPolicy;
+import uk.nhs.adaptors.gp2gp.common.exception.RetryLimitReachedException;
+import uk.nhs.adaptors.gp2gp.gpc.configuration.GpcConfiguration;
+import uk.nhs.adaptors.gp2gp.gpc.exception.GpcServerErrorException;
+import uk.nhs.adaptors.gp2gp.testcontainers.MongoDBExtension;
+
+@SpringBootTest
+@ExtendWith({SpringExtension.class, MockitoExtension.class, MongoDBExtension.class})
+@DirtiesContext
+public class GpcWebClientTest {
+    private static final String DOCUMENT_ID = "testDocumentId";
+    private static final String MESSAGE_ID = "testMessageId";
+    private static final String FROM_ASID = "testFromAsid";
+    private static final String TO_ASID = "testToAsid";
+    private static final String FROM_ODS = "testFromOdsCode";
+    private static final String TO_ODS = "testToOdsCode";
+    private static final String CONVERSATION_ID = "testConversationId";
+    private static final String REQUEST_ID = "testRequestId";
+    private static final String TASK_ID = "testTaskId";
+    private static final String TEST_BODY = "Test Body";
+    private static final int FOUR = 4;
+
+    private static MockWebServer mockWebServer;
+
+    private String baseUrl;
+    @SpyBean
+    private GpcConfiguration gpcConfiguration;
+    @Autowired
+    private GpcClient gpcWebClient;
+
+    @BeforeEach
+    public void initialise() throws IOException {
+        mockWebServer = new MockWebServer();
+        mockWebServer.start();
+        baseUrl = String.format("http://localhost:%s", mockWebServer.getPort());
+        when(gpcConfiguration.getUrl()).thenReturn(baseUrl);
+    }
+
+    @AfterEach
+    public void tearDown() throws IOException {
+        mockWebServer.shutdown();
+    }
+
+    @Test
+    public void When_GetDocumentRecord_With_HttpStatus200_Expect_NoException() {
+        MockResponse response = new MockResponse();
+        response.setResponseCode(OK.value())
+            .setBody(TEST_BODY);
+
+        mockWebServer.enqueue(response);
+
+        var taskDefinition = buildDocumentTaskDefinition();
+        var result = gpcWebClient.getDocumentRecord(taskDefinition);
+
+        assertThat(result).isEqualTo(TEST_BODY);
+    }
+
+    @Test
+    public void When_GetDocumentRecord_With_HttpStatus5xx_Expect_RetryExceptionWithGpcServerErrorExceptionAsRootCause() {
+        MockResponse response = new MockResponse();
+        response.setResponseCode(INTERNAL_SERVER_ERROR.value());
+        response.setBody(TEST_BODY);
+
+        mockWebServer.enqueue(response);
+        mockWebServer.enqueue(response);
+        mockWebServer.enqueue(response);
+        mockWebServer.enqueue(response);
```

Comment:
🤣 Hmmmm.

Reply:
Haha, the library doesn't seem to give another way to enqueue responses and I wanted to avoid putting a loop in the test. But can do it if you feel it's preferable to the duplication?

Reply:
I've refactored it using a loop